doc/spec/namespace.md

```diff
@@ -0,0 +1,159 @@
+# Namespace
+The namespace represents the full set of named repositories which can be
+referenced by Docker. A repository contains a set of content-addressable blobs
+and tags referencing those blobs. The namespace can be used to discover the
+location and trust information of a repository based on DNS, HTTP requests and
+other means. The namespace is global and owned through the domain at the root of
+the namespace path.
+
+A repository should always be referenced by its fully canonical value. If an
+engine presents a shortened name to a user, that name should be fully
+canonicalized based on rules defined by the engine before contacting a remote
+server. When mirroring repositories, the original name for a repository must be
+used. Changing a repository name is equivalent to copying or moving and will not
+be referenced by the original repository.
+
+## Discovery
+The discovery process involves resolving a namespace into registry metadata. The
+registry metadata contains the full set of information needed to fetch and
+verify content associated with the namespace repository. The metadata includes
+list of registry endpoints, the trust model, and search index.
+
+Discovery can be defined as...
+`<namespace> -> ([<endpoint>, ...], <trust info>, <search index>)`
+
+or in Go as...
+~~~go
+type RegistryResolver interface {
+	Resolve(name string) Registry
+}
+~~~
+
+
+### Resolve Methods
+
+The first element of the namespace is extracted and used as the domain name for
+resolution. The domain name should be used unmodified.
+
+`<namespace> -> <domain>/<name>`
+
+#### HTTPS
+The discovery related metadata will be fetched via HTTPS from the DNS resolved
+location using the remaining namespace path elements as the HTTP path in a GET
+request. A discovery request url would be in the format
+https://<domain>/<name>?docker-discovery=1
+
+For example, “example.com/foo/bar” would create a url
+“https://example.com/foo/bar?docker-discovery=1”
+
+##### HTML Response
+~~~html
+<meta name="docker-scope" content="registry.example.com">
+<meta name=“docker-registry” content=“push,pull v2 https://registry.example.com/v2/”>
+<meta name=“docker-registry” content=“push,pull v1 https://registry.example.com/v1/”>
+<meta name=“docker-registry” content=“pull v2 https://registry.mirror.com/v2/”>
+<meta name=“docker-registry” content=“pull,notag v2 http://registry.mirror.com/v2/”>
+<meta name=“docker-search” content=“v1 https://search.mirror.com/v1/”>
+<meta name=“docker-trust” content=“gpg https://keys.example.com/{name}.key”>
+<meta name=“docker-trust” content=“tuf https://tuf.example.com/{name}/}”>
+~~~
+
+##### JSON Response
+~~~json
```

Comment:
meta tags give the flexibility of a dumb website managed via jekyll or a static site to add discovery without having to support content/type switching or parsing of the HTTP GET variables. The compromise seemed reasonable since a large number of interesting community driven projects use simple web hosting like github.io, gs, or s3.